File: Python/Set.py
# initialize set
# NOTE: set of integers may appear to be sorted because integers usually hash to itself.
setA = {1, 2, 4, 3, 6, 9, 4, 6, 2}
setB = set((2, 5, 4, 7, 9, 4, 2))
setC = {"hello", "animal", "dog", "coffee", "donut", "hi"}
print(setA)
print(setB)
# for string, unordered trait is visible because hashing is not itself
print(setC)

# Set Operations
# intersection can be done using "&" operator as well.
intersection = setA.intersection(setB)
print("Intersection: " + str(intersection))
intersection = setA & setB
print("Intersection: " + str(intersection))

# union can be done using "|" operator as well.
union = setA.union(setB)
print("Union: " + str(union))
union = setA | setB
print("Union: " + str(union))

# difference can be done using "-" operator as well.
difference = setA.difference(setB)
print("Difference: " + str(difference))
difference = setA - setB
print("Difference: " + str(difference))

# symmetric difference = union - intersection (in either, but not in both)
# symmetric difference can be done using "^" operator as well.
symDifference = setA.symmetric_difference(setB)
print("Symmetric Difference: " + str(symDifference))
symDifference = setA ^ setB
print("Symmetric Difference: " + str(symDifference))

File: Javascript/Set.js
// initialize Set
let setA = new Set([1, 4, 3, 7, 5, 4, 5, 3]);
let setB = new Set([2, 6, 3, 7, 5, 3]);
let setC = new Set([5, 3, 3, 7, 5, 5, 1]);
console.log("Set A:", setA);
console.log("Set B:", setB);
console.log("Set C:", setC);

console.log(isSuperset(setA, setC));          // => true
console.log(isSuperset(setB, setC));          // => false
console.log(union(setA, setB));               // => Set [1, 4, 3, 7, 5, 2, 6]
console.log(intersection(setA, setB));        // => Set [3, 7, 5]
console.log(symmetricDifference(setA, setB)); // => Set [1, 4, 2, 6]
console.log(difference(setA, setB));          // => Set [1, 4]

// Basic Set Operations
// source: MDN doc (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)
function isSuperset(set, subset) {
    for (let elem of subset) {
        if (!set.has(elem)) {
            return false
        }
    }
    return true
}

function union(setA, setB) {
    let _union = new Set(setA)
    for (let elem of setB) {
        _union.add(elem)
    }
    return _union
}

function intersection(setA, setB) {
    let _intersection = new Set()
    for (let elem of setB) {
        if (setA.has(elem)) {
            _intersection.add(elem)
        }
    }
    return _intersection
}

function symmetricDifference(setA, setB) {
    let _difference = new Set(setA)
    for (let elem of setB) {
        if (_difference.has(elem)) {
            _difference.delete(elem)
        } else {
            _difference.add(elem)
        }
    }
    return _difference
}

function difference(setA, setB) {
    let _difference = new Set(setA)
    for (let elem of setB) {
        _difference.delete(elem)
    }
    return _difference
}